packages/extension/src/companion/App.tsx

```diff
@@ -0,0 +1,251 @@
+import MenuIcon from '@dailydotdev/shared/icons/menu.svg';
+import ArrowIcon from '@dailydotdev/shared/icons/arrow.svg';
+import LogoIcon from '@dailydotdev/shared/src/svg/LogoIcon';
+import {
+  HotLabel,
+  TLDRText,
+} from '@dailydotdev/shared/src/components/utilities';
+import classNames from 'classnames';
+import SimpleTooltip from '@dailydotdev/shared/src/components/tooltips/SimpleTooltip';
+import { QueryClient, QueryClientProvider } from 'react-query';
+import Modal from 'react-modal';
+import { useContextMenu } from '@dailydotdev/react-contexify';
+import useNotification from '@dailydotdev/shared/src/hooks/useNotification';
+import { CardNotification } from '@dailydotdev/shared/src/components/cards/Card';
+import { isTesting } from '@dailydotdev/shared/src/lib/constants';
+import CompanionContextMenu from './CompanionContextMenu';
+import { BootData } from './common';
+import '@dailydotdev/shared/src/styles/globals.css';
+import useCompanionActions from './useCompanionActions';
+import { browser } from 'webextension-polyfill-ts';
+
+const queryClient = new QueryClient();
+
+if (!isTesting) {
+  Modal.setAppElement('daily-companion-app');
+}
+
+function InternalApp({ postData, onOptOut }: { postData: BootData; onOptOut }) {
+  const [post, setPost] = useState<BootData>(postData);
+  const [companionState, setCompanionState] = useState<boolean>(false);
+  const { notification, onMessage } = useNotification();
+
+  const firstLoad = useRef(false);
+  useEffect(() => {
+    firstLoad.current = true;
+  }, []);
+
+  const updatePost = async (update) => {
+    const oldPost = post;
+    setPost({
+      ...post,
+      ...update,
+    });
+    return () => setPost(oldPost);
+  };
+
+  const {
+    bookmark,
+    removeBookmark,
+    upvote,
+    removeUpvote,
+    report,
+    blockSource,
+    disableCompanion,
+  } = useCompanionActions({
+    onBookmarkMutate: () => updatePost({ bookmarked: true }),
+    onRemoveBookmarkMutate: () => updatePost({ bookmarked: false }),
+    onUpvoteMutate: () =>
+      updatePost({ upvoted: true, numUpvotes: post.numUpvotes + 1 }),
+    onRemoveUpvoteMutate: () =>
+      updatePost({ upvoted: false, numUpvotes: post.numUpvotes + -1 }),
+  });
+
+  const optOut = () => {
+    disableCompanion({});
+    onOptOut();
+  };
+
+  const toggleUpvote = async () => {
+    if (!post.upvoted) {
+      await upvote({ id: post.id });
+    } else {
+      await removeUpvote({ id: post.id });
+    }
+  };
+
+  const toggleBookmark = async () => {
+    if (!post.bookmarked) {
+      await bookmark({ id: post.id });
+    } else {
+      await removeBookmark({ id: post.id });
+    }
+  };
+
+  const { show: showCompanionOptionsMenu } = useContextMenu({
+    id: 'companion-options-context',
+  });
+  const onContextOptions = (event: React.MouseEvent): void => {
+    showCompanionOptionsMenu(event, {
+      position: { x: 48, y: 265 },
+    });
+  };
+
+  return (
+    <div
```

Comment:
Perhaps we can refactor this into multiple smaller components

Reply:
I agree this is too big and hard to read